### PR TITLE
[SR-4685] StarRocksColumn convert to Arrow Array

### PR DESCRIPTION
**Pull request by satanson**
## Changes
Add vectorized method (send_chunk) for MemoryScratchSink to convert StarRocks in-memory column into Arrow array.

### CODE REVIEW

File: be/src/util/arrow/starrocks_column_to_arrow.cpp [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+IS_CONV_FLOAT_INTEGER(TYPE_TINYINT, ArrowTypeId::INT8)
+IS_CONV_FLOAT_INTEGER(TYPE_SMALLINT, ArrowTypeId::INT16)
+IS_CONV_FLOAT_INTEGER(TYPE_INT, ArrowTypeId::INT32)
+IS_CONV_FLOAT_INTEGER(TYPE_BIGINT, ArrowTypeId::INT64)
+IS_CONV_FLOAT_INTEGER(TYPE_FLOAT, ArrowTypeId::FLOAT)
+IS_CONV_FLOAT_INTEGER(TYPE_DOUBLE, ArrowTypeId::DOUBLE)
+IS_CONV_FLOAT_INTEGER(TYPE_TIME, ArrowTypeId::DOUBLE)
+
+template <PrimitiveType PT, ArrowTypeId AT, bool is_nullable>
+struct ColumnToArrowConverter<PT, AT, is_nullable, ConvFloatAndIntegerGuard<PT, AT>> {
```
Comment (imay) [inline be/src/util/arrow/starrocks_column_to_arrow.cpp:29]: should add some UT

**Comment by satanson** [comment]
run starrocks_tscancode

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by satanson** [comment]
run starrocks_clang-tidy

**Comment by satanson** [comment]
run starrocks_clang-format

**Comment by satanson** [comment]
run starrocks_be_unittest

**Comment by satanson** [comment]
run starrocks_tscancode

File: be/src/runtime/primitive_type.h [hunk 2]
```diff
@@ -212,11 +213,13 @@ VALUE_GUARD(PrimitiveType, Decimal64PTGuard, pt_is_decimal64, TYPE_DECIMAL64)
 VALUE_GUARD(PrimitiveType, Decimal128PTGuard, pt_is_decimal128, TYPE_DECIMAL128)
 VALUE_GUARD(PrimitiveType, DecimalPTGuard, pt_is_decimal, TYPE_DECIMAL32, TYPE_DECIMAL64, TYPE_DECIMAL128)
 VALUE_GUARD(PrimitiveType, SumDecimal64PTGuard, pt_is_sum_decimal64, TYPE_DECIMAL32, TYPE_DECIMAL64)
+VALUE_GUARD(PrimitiveType, HllPTGuard, pt_is_hll, TYPE_HLL)
 VALUE_GUARD(PrimitiveType, BinaryPTGuard, pt_is_binary, TYPE_CHAR, TYPE_VARCHAR)
 
 VALUE_GUARD(PrimitiveType, DatePTGuard, pt_is_date, TYPE_DATE)
 VALUE_GUARD(PrimitiveType, DateTimePTGuard, pt_is_datetime, TYPE_DATETIME)
 VALUE_GUARD(PrimitiveType, TimePTGuard, pt_is_time, TYPE_TIME)
+VALUE_GUARD(PrimitiveType, DecimalV1PTGuard, pt_is_decimalv1, TYPE_DECIMAL)
```
Comment (kangkaisen) [inline be/src/runtime/primitive_type.h:222]: Needn't support decimal v1. We will remove decimal v1 related codes soon

File: be/src/util/arrow/starrocks_column_to_arrow.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#include <memory>
+
+#include "column/chunk.h"
+#include "common/global_types.h"
+#include "runtime/descriptors.h"
+
+namespace starrocks {
+namespace vectorized {
+
+Status vectorized_convert_to_arrow_batch(Chunk* chunk, const std::vector<const TypeDescriptor*>& _slot_types,
```
Comment (kangkaisen) [inline be/src/util/arrow/starrocks_column_to_arrow.h:26]:
```suggestion
Status convert_chunk_to_arrow_batch(Chunk* chunk, const std::vector<const TypeDescriptor*>& _slot_types,
```

File: be/src/util/arrow/starrocks_column_to_arrow.cpp [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+                                         std::shared_ptr<arrow::RecordBatch>* result) {
+    if (chunk->num_columns() != schema->num_fields()) {
+        return Status::InvalidArgument("number fields not match");
+    }
+
+    std::vector<std::shared_ptr<arrow::Array>> arrays(slot_types.size());
+
+    for (auto i = 0; i < slot_types.size(); ++i) {
+        auto column = chunk->get_column_by_slot_id(slot_ids[i]);
+        if (column->only_null() || column->is_constant()) {
```
Comment (kangkaisen) [inline be/src/util/arrow/starrocks_column_to_arrow.cpp:295]: check `column->is_constant()` is enough.

File: be/src/runtime/memory_scratch_sink.cpp [hunk 5]
```diff
@@ -48,13 +49,24 @@ MemoryScratchSink::MemoryScratchSink(const RowDescriptor& row_desc, const std::v
...
+}
+
 Status MemoryScratchSink::prepare_exprs(RuntimeState* state) {
     // From the thrift expressions create the real exprs.
     RETURN_IF_ERROR(Expr::create_expr_trees(state->obj_pool(), _t_output_expr, &_output_expr_ctxs));
     // Prepare the exprs to run.
     RETURN_IF_ERROR(Expr::prepare(_output_expr_ctxs, state, _row_desc, _expr_mem_tracker.get()));
     // generate the arrow schema
     RETURN_IF_ERROR(convert_to_arrow_schema(_row_desc, &_arrow_schema));
+    convert_to_slot_types_and_ids();
```
Comment (kangkaisen) [inline be/src/runtime/memory_scratch_sink.cpp:69]: RETURN_IF_ERROR

File: be/src/runtime/primitive_type.h (hunk 2 above)
Comment (satanson) [reply be/src/runtime/primitive_type.h:222]: ok

File: be/src/util/arrow/starrocks_column_to_arrow.h (hunk 3 above)
Comment (satanson) [reply be/src/util/arrow/starrocks_column_to_arrow.h:26]: ok

File: be/src/util/arrow/starrocks_column_to_arrow.cpp (hunk 4 above)
Comment (satanson) [reply be/src/util/arrow/starrocks_column_to_arrow.cpp:295]: ok

File: be/src/runtime/memory_scratch_sink.cpp (hunk 5 above)
Comment (satanson) [reply be/src/runtime/memory_scratch_sink.cpp:69]: it always succeeds and return type is void

**Review by kangkaisen — APPROVED** [review]

File: be/test/CMakeLists.txt [hunk 6]
```diff
@@ -279,3 +279,7 @@ TARGET_LINK_LIBRARIES(bytes_test
         gtest
         tcmalloc
         )
+
+add_executable(starrocks_column_to_arrow_test ./util/arrow/starrocks_column_to_arrow_test.cpp)
+TARGET_LINK_LIBRARIES(starrocks_column_to_arrow_test ${TEST_LINK_LIBS})
+SET_TARGET_PROPERTIES(starrocks_column_to_arrow_test PROPERTIES COMPILE_FLAGS "-fno-access-control")
```
Comment (dirtysalt) [inline be/test/CMakeLists.txt:285]: is that compiler flag for testing?

Comment (satanson) [reply be/test/CMakeLists.txt:285]: building a huge starrocks_test file takes so long time

File: be/src/util/arrow/starrocks_column_to_arrow.cpp [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    static inline arrow::Decimal128 convert_datum(const StarRocksCppType& datum) {
+        int128_t value;
+        if constexpr (pt_is_decimalv2<PT>) {
+            value = datum.value();
+        } else if constexpr (pt_is_decimal<PT>) {
+            value = datum;
+        } else {
+            static_assert(pt_is_decimalv2<PT> || pt_is_decimal<PT>, "Illegal PrimitiveType");
+        }
+        int64_t high = value >> 64;
```
Comment (dirtysalt) [inline be/src/util/arrow/starrocks_column_to_arrow.cpp:85]: AFAIK, someone in the community is working on porting StarRocks to ARM. Should we care about little endian and big endian from now?

**Review by dirtysalt — APPROVED** [review]

File: be/src/util/arrow/starrocks_column_to_arrow.cpp (hunk 7 above)
Comment (satanson) [reply be/src/util/arrow/starrocks_column_to_arrow.cpp:85]: no need, shift operations are immune

**Pull request merged**